storage/innobase/fil/fil0fil.cc

```diff
@@ -10453,6 +10453,30 @@ bool Fil_system::open_for_recovery(space_id_t space_id) {
       fil_tablespace_encryption_init(space);
     }
 
+    if (recv_sys->crypt_datas != nullptr &&
+        recv_sys->crypt_datas->count(space_id) > 0) {
+      if (space->crypt_data != nullptr) {
+        fil_space_destroy_crypt_data(&space->crypt_data);
+      }
+      space->crypt_data = (*recv_sys->crypt_datas)[space_id];
+      recv_sys->crypt_datas->erase(space_id);
```

Comment:
This is a bit tricky area. I think we should trust crypt_data found in page 0? Looks like, here you destroy the crypt_data found from page 0 read and use the one from recovery. IMHO, we should use crypt_data from recovery only if there is no crypt_data found in page 0.

Am also bit unsure on when to use which crypt_data (from the one recovery or from page 0)... really depends on ALTERs.. Can you please give some more thought? see if it covers the ALTER scenarios as well.

Reply:
The write of crypt_data to page0 is always creating a full redo log record of crypt_data. Based on this record we recreate the whole crypt_data structure. Whatever is there in a redo - this is the newest (full) version of crypt_data and should replace existing crypt_data.

Reply:
Not always. what if there is cryp_data from redo and the current table is unencrypted? i.e sequence of ALTERs with final once as unencrypted. I guess you are in the safe zone because you need table rebuilding ALTER to make it unencrypted? :thinking: 

So what would happen is, redo parsing will use crypt_data for tablespace, but in reality, the table should be unencrypted?

Reply:
I have added Encryption_rotation::ENCRYPTING to address this issue. Please check the commit message.